client_test.go

```diff
@@ -0,0 +1,65 @@
+package metricd
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+)
+
+// Should be run in a separate goroutine. Runs a grpc server at the
+// given port after sleeping for 3 seconds. The listener will be passed
+// to listenerChan, or listenerChan will be closed if there was an error.
+func runSlowListeningServer(t *testing.T, port int, listenerChan chan net.Listener) {
+	time.Sleep(3 * time.Second)
+	server := grpc.NewServer()
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		fmt.Println(err)
+		close(listenerChan)
+		return
+	}
+	// Let the main goroutine know about the listener so they can close it
+	listenerChan <- l
+	server.Serve(l)
+}
+
+func TestClientConnect(t *testing.T) {
+	// Set up a dummy grpc server that takes a few seconds to
+	// start listening
+	const port = 6789
+	listenerChan := make(chan net.Listener, 1)
+	go runSlowListeningServer(t, port, listenerChan)
+	defer func() {
+		l, success := <-listenerChan
+		if success {
+			l.Close()
+		}
+	}()
+
+	// Create a client
+	url := url.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("localhost:%d", port),
+	}
+	client, err := NewClient(&url)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	// Try connecting
+	rc := client.(*realClient)
+	err = rc.tryConnect()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if rc.conn.State() != grpc.Ready {
```

Comment:
I agree that a race is technically possible here, but given the time scale involved in timing out a network connection, I'd argue that it is orders of magnitude less likely to happen compared to the flaky tests we've seen in the past. Do you know of a strategy for testing this sort of network-based thing that is time-bounded and not potentially flaky?

But yeah I'll certainly add a comment.